### PR TITLE
test: Disable shadow catalog in parallel tests

### LINKED ISSUE #23146: parallel-workload: thread 'main' panicked at 'assertion failed: `(left == right)`
### What version of Materialize are you using?

1e19b74517

### What is the issue?

Seen in nightly during the weekend: https://buildkite.com/materialize/nightlies/builds/5150#018bc383-5add-4929-821b-55b1b4d3a09c
I'm not sure if this is a regression, since the failure is flaky:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Transaction { ... }`
 right: `Transaction { ... }`
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at ./rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:67:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed::<mz_catalog::durable::transaction::Transaction, mz_catalog::durable::transaction::Transaction>
             at ./rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:228:5
   4: <mz_catalog::durable::impls::shadow::ShadowCatalogState as mz_catalog::durable::DurableCatalogState>::transaction::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/catalog/src/durable/impls/shadow.rs:202:44
   5: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = core::result::Result<mz_catalog::durable::transaction::Transaction, mz_catalog::durable::error::CatalogError>> + core::marker::Send>> as core::future::future::Future>::poll
             at ./rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/future/future.rs:125:9
   6: <mz_adapter::catalog::Catalog>::open::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/adapter/src/catalog/open.rs:258:45
   7: <tracing::instrument::Instrumented<<mz_adapter::catalog::Catalog>::open::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
   8: <mz_adapter::catalog::Catalog>::open::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/adapter/src/catalog/open.rs:186:5
   9: mz_adapter::coord::serve::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/adapter/src/coord.rs:2078:10
  10: <mz_environmentd::Listeners>::serve::{closure#0}::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/environmentd/src/lib.rs:567:10
  11: <tracing::instrument::Instrumented<<mz_environmentd::Listeners>::serve::{closure#0}::{closure#0}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  12: <mz_environmentd::Listeners>::serve::{closure#0}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/environmentd/src/lib.rs:339:5
  13: environmentd::run::{closure#11}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/environmentd/src/bin/environmentd/main.rs:1000:14
  14: <tracing::instrument::Instrumented<environmentd::run::{closure#11}> as core::future::future::Future>::poll
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  15: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<environmentd::run::{closure#11}>>::{closure#0}
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/park.rs:282:63
  16: tokio::runtime::coop::with_budget::<core::task::poll::Poll<core::result::Result<mz_environmentd::Server, anyhow::Error>>, <tokio::runtime::park::CachedParkThread>::block_on<tracing::instrument::Instrumented<environmentd::run::{closure#11}>>::{closure#0}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/coop.rs:107:5
  17: tokio::runtime::coop::budget::<core::task::poll::Poll<core::result::Result<mz_environmentd::Server, anyhow::Error>>, <tokio::runtime::park::CachedParkThread>::block_on<tracing::instrument::Instrumented<environmentd::run::{closure#11}>>::{closure#0}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/coop.rs:73:5
  18: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<environmentd::run::{closure#11}>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/park.rs:282:31
  19: <tokio::runtime::context::blocking::BlockingRegionGuard>::block_on::<tracing::instrument::Instrumented<environmentd::run::{closure#11}>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context/blocking.rs:66:9
  20: <tokio::runtime::scheduler::multi_thread::MultiThread>::block_on::<tracing::instrument::Instrumented<environmentd::run::{closure#11}>>::{closure#0}
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/mod.rs:87:13
  21: tokio::runtime::context::runtime::enter_runtime::<<tokio::runtime::scheduler::multi_thread::MultiThread>::block_on<tracing::instrument::Instrumented<environmentd::run::{closure#11}>>::{closure#0}, core::result::Result<mz_environmentd::Server, anyhow::Error>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/context/runtime.rs:65:16
  22: <tokio::runtime::scheduler::multi_thread::MultiThread>::block_on::<tracing::instrument::Instrumented<environmentd::run::{closure#11}>>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/scheduler/multi_thread/mod.rs:86:9
  23: <tokio::runtime::runtime::Runtime>::block_on::<environmentd::run::{closure#11}>
             at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.32.0/src/runtime/runtime.rs:349:50
  24: environmentd::run
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/environmentd/src/bin/environmentd/main.rs:930:18
  25: environmentd::main
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-017175357bd56887b-1/materialize/tests/src/environmentd/src/bin/environmentd/main.rs:625:23
  26: <fn() as core::ops::function::FnOnce<()>>::call_once
             at ./rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5
```
Full assertion failure is huge: https://gist.github.com/def-/6e38fe91fa412524d752d66fe8575e11
Using `diff` the replica next_id is different:
```diff
22821c22821
< IdAllocKey { name: "replica" }: IdAllocValue { next_id: 141 },
---
> IdAllocKey { name: "replica" }: IdAllocValue { next_id: 140 },
```
I believe this might be related to the shadow catalog introduced in https://github.com/MaterializeInc/materialize/pull/23086

### PR DESCRIPTION
**Pull request by jkosh44**
The Shadow catalog does the following on every catalog operation:

  1. Execute the operation on catalog A.
  2. Execute the operation on catalog B.
  3. Compare the results.

The parallel-workload tests will randomly kill Materialize, so we could end up in a situation where the operation happens on catalog A but not on catalog B, and we will see an inconsistency between the two catalogs. The shadow catalog is meant to be used to compare two catalog implementations, so it's OK that it has this deficiency. Given that, this commit disables the shadow catalog in parallel-workload tests.

Fixes #23146

### Motivation
This PR fixes a recognized bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Review by def- — CHANGES_REQUESTED** [review]
We only kill materialized randomly in the KillScenario. Would it make sense to keep it for the other scenarios? Probably needs a Materialized-override then for the c.up in line 68 based on the value of `args.scenario == Scenario.Kill`.

**Comment by jkosh44** [comment]
> We only kill materialized randomly in the KillScenario. Would it make sense to keep it for the other scenarios? Probably needs a Materialized-override then for the c.up in line 68 based on the value of `args.scenario == Scenario.Kill`.

If possible then yes, it would make sense, however I have no idea how the parallel tests work. The one constraint is that once Materialize boots up with one catalog store, it must use always use the same catalog store.

**Comment by def-** [comment]
I'll push into your branch.

**Comment by def-** [comment]
I'm afraid there might be more tests with this problem, let's see:
```
balancerd/mzcompose.py:    c.kill("materialized")
balancerd/mzcompose.py:    c.kill("materialized")
cluster/mzcompose.py:    c.kill("materialized")
cluster/mzcompose.py:        c.kill("materialized")
cluster/mzcompose.py:        c.kill("materialized")
cluster/mzcompose.py:        c.kill("materialized")
cluster/mzcompose.py:        c.kill("materialized")
cluster/mzcompose.py:        c.kill("materialized")
cluster/mzcompose.py:    c.kill("materialized")
cluster/mzcompose.py:    c.kill("materialized")
cluster/mzcompose.py:    c.kill("materialized")
incident-72/mzcompose.py:        c.kill("materialized")
incident-72/mzcompose.py:    c.kill("materialized")
incident-72/mzcompose.py:        c.kill("materialized")
kafka-exactly-once/mzcompose.py:    c.kill("materialized")
kafka-ingest-open-loop/mzcompose.py:    c.kill("materialized")
legacy-upgrade/mzcompose.py:    c.kill("materialized")
legacy-upgrade/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:    c.kill("materialized")
persistence/mzcompose.py:        c.kill("materialized")
pg-cdc-resumption/mzcompose.py:    c.kill("materialized")
replica-isolation/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
restart/mzcompose.py:    c.kill("materialized")
ssh-connection/mzcompose.py:    c.kill("materialized")
upsert/mzcompose.py:            c.kill("materialized")
upsert/mzcompose.py:            c.kill("materialized")
```

**Comment by def-** [comment]
I'm pretty confused as to why I'm getting new errors in kill scenario now with stash catalog: https://buildkite.com/materialize/nightlies/builds/5180#018bcb17-434b-4cd9-a971-baa246182f32
Did something change? Haven't seen those before, otherwise I'll investigate on Wednesday (I'm out on Tuesday)

**Comment by jkosh44** [comment]
> I'm pretty confused as to why I'm getting new errors in kill scenario now with stash catalog: https://buildkite.com/materialize/nightlies/builds/5180#018bcb17-434b-4cd9-a971-baa246182f32 Did something change? Haven't seen those before, otherwise I'll investigate on Wednesday (I'm out on Tuesday)

I don't think anything changed. What is the error? I'm having a hard time finding it in the test output. Is it this?

```
Exception in thread worker_30:
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/parallel_workload/worker.py", line 87, in run
    ReconnectAction(self.rng, self.composition, random_role=False).run(
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/parallel_workload/action.py", line 903, in run
    autocommit = exe.cur._c.autocommit
                 ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'autocommit'
Exception in thread worker_31:
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/parallel_workload/worker.py", line 87, in run
    ReconnectAction(self.rng, self.composition, random_role=False).run(
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/parallel_workload/action.py", line 903, in run
    autocommit = exe.cur._c.autocommit
                 ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'autocommit'
Exception in thread worker_21:
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/parallel_workload/worker.py", line 87, in run
    ReconnectAction(self.rng, self.composition, random_role=False).run(
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/parallel_workload/action.py", line 903, in run
    autocommit = exe.cur._c.autocommit
                 ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'autocommit'
QPS: 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.1 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.0 000.1 000.1 000.0
mzcompose: test case workflow-default failed: <class 'Exception'>: Thread worker_21 failed, exiting
Traceback (most recent call last):
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/mzcompose/composition.py", line 470, in test_case
    yield
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/cli/mzcompose.py", line 591, in handle_composition
    composition.workflow(args.workflow, *args.unknown_subargs[1:])
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/mzcompose/composition.py", line 360, in workflow
    func(self, parser)
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/test/parallel-workload/mzcompose.py", line 94, in workflow_default
    run(
  File "/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0cf3b7dfef0a2803e-1/materialize/nightlies/misc/python/materialize/parallel_workload/parallel_workload.py", line 274, in run
    raise Exception(f"Thread {thread.name} failed, exiting")
Exception: Thread worker_21 failed, exiting
--- Uploading report for suite 'mzcompose' to Buildkite Test Analytics
202 {"id":"8e5670ea-55d0-490d-b810-2b15604248bb","run_id":"c5daa355-148b-4ec1-877d-fc1b964aeb00","queued":1,"skipped":0,"errors":[],"run_url":"https://buildkite.com/organizations/materialize/analytics/suites/mzcompose/runs/c5daa355-148b-4ec1-877d-fc1b964aeb00"}
mzcompose: error: at least one test case failed
```

If so, that looks like an issue in python interacting with the PostgreSQL python library, and not something related to the catalog.

**Review by philip-stoev — DISMISSED** [review]
The left == right assertion also happens in Platform Checks backup + restore, the diff there is in one timestamp value. I think all of the platform checks need to be switched to the safe implementation, as almost all of them will kill Mz at some point or another.

**Review by def- — APPROVED** [review]
Nightly looks good now: https://buildkite.com/materialize/nightlies/builds/5215
Can be submitted imo (but I now wrote the code,so will wait for another review.

**Review by nrainer-materialize — APPROVED** [review]
Fine with me if nightly is happy.

**Comment by def-** [comment]
This decreases nightly flakiness. Hopefully we have all the interesting tests excluded now.

**Comment by jkosh44** [comment]
> The left == right assertion also happens in Platform Checks backup + restore, the diff there is in one timestamp value. I think all of the platform checks need to be switched to the safe implementation, as almost all of them will kill Mz at some point or another.

This just made me realize that any test that shuts down Materialize and then brings it back up has the chance to flake due to timestamps. Even an idle Materialize will update the timestamps every 10 seconds. If we happen to restart in the middle of that a timestamp update, it's possible that only one of the catalog gets updated. Are there other tests where we do that?

**Comment by def-** [comment]
Nearly every test does that. Could we change the assertion not to compare timestamps for that but everything else?

**Comment by jkosh44** [comment]
> Nearly every test does that. Could we change the assertion not to compare timestamps for that but everything else?

Yeah, I'll put something together today.

**Comment by jkosh44** [comment]
> > Nearly every test does that. Could we change the assertion not to compare timestamps for that but everything else?
> 
> Yeah, I'll put something together today.

https://github.com/MaterializeInc/materialize/pull/23219

**Pull request merged**
